### PR TITLE
fix: Increase line buffer to avoid frequent interleaved logs

### LINKED ISSUE #1806: Regression of interleaved logs
We fixed #1615 by using a `LineWriter`, so that each thread buffers a full line before writing and flushing.
However, turns out that `LineWriter` has a default buffer size of 1024, and when that is exceeded it causes multiple writer to the underlying writer (stdout).

Once we merged #1792, we effectively produced many interleaved log lines
This is an example in the current output of a router iox role in our staging env:

```console
$ grep 'level=' /tmp/logs | egrep -v '^level=' | ag level=debug
1:"Request { method: POST, uri: /api/v2/write?bucket=eaec8df57a81a1e9&org=844910ece80be8bc, version: HTTP/1.1, headers: {\"host\": \"influxdb-istio.aws.influxdata.io-shadow\", \"user-agent\": \"Telegraf/1.14.5\", \"content-encoding\": \"gzip\", \"content-type\": \"text/plain; charset=utf-8\", \"transfer-encoding\": \"chunked\", \"accept-encoding\": \"gzip\", \"x-forwarded-for\": \"10.200.5.108,10.141.66.33\", \"x-forwarded-proto\": \"https\", \"x-request-id\": \"e8abff66-522b-45d7-a87b-c169f9c699ec\", \"x-envoy-attempt-count\": \"1\", \"x-forwarded-client-cert\": \"By=spiffe://toolsus1.a.influxcloud.net/ns/iox/sa/iox;Hash=e4c867284b07b50862e4cbfa59e618787f268fb69340ea696c80720404dd88f9;Subject=\\\"\\\";URI=spiffe://toolsus1.a.influxcloud.net/ns/istio-system/sa/ingressgateway-private-service-account\", \"x-b3-traceid\": \"7de068974c0e8fde53040db5d1a4f706\", \"x-b3-spanid\": \"a3a00520df33bcaf\", \"x-b3-parentspanid\": \"dca6fa8583346086\", \"x-b3-sampled\": \"0\"}, body: Body(Streaming) }" span=level=debug msg="Successfully processed request" response="Response { status: 500, version: HTTP/1.1, headers: {\"x-powered-by\": \"Routerify v2.0.0-beta-2\"}, body: Body(Full(b\"{\\\"error\\\":\\\"Internal error writing points into org 844910ece80be8bc, bucket 64881f0f8fd1653c:  remote error: cannot write to remote: Unexpected server error: The service is currently unavailable: no healthy upstream\\\",\\\"error_code\\\":100}\")) }" target="influxdb_iox::influxdb_ioxd::http" location="src/influxdb_ioxd/http.rs:356" time=1624620941815970656
```

notice the `span=level=debug msg....` in the middle of this line.

This in turn caused a cascading failure in our graylog log indexing system:

1.
```
2021-06-24 19:26:29,805 WARN : org.graylog2.indexer.messages.Messages - Failed to index message: index=<teamconductor_1477> id=<ef7e490b-d521-11eb-83a3-b68e1618e7df> error=<{"type":"illegal_argument_exception","reason":"Limit of total fields [1000] in index [teamconductor_1477] has been exceeded"}>
```

2. the failed indexing caused a buildup in the (global) graylog journal which forced us to rollback a release because we were breaking the logging infrastructure for everybody else.

### PR DESCRIPTION
**Pull request by mkmik**
This is a horrible hack to fix #1806

We're currently corrupting a measurable fraction of our logs, and we disabled the verbose debug logs, so we're not causing immediate grief to graylog, but still.

```console
$ kubectl logs deploy/iox-router-1001 -c iox | tee /tmp/logs \
  | awk -F 'level=' '{print (NF?NF-1:0)}' \
  | sort | uniq -c
 234 0
57422 1
 230 2
   2 3
```

---

The longer term plan for this is still to be defined. A new release of the `tracing-subscriber` crate would allow a better fix. We could depend on the unreleased crate, or do something else.

### CODE REVIEW

File: trogging/src/lib.rs [hunk 1]
```diff
@@ -411,6 +429,31 @@ impl Drop for TracingGuard {
...
+{
+    fmt::writer::BoxMakeWriter::new(move || {
+        LimitedWriter(
+            MAX_LINE_LENGTH,
+            std::io::LineWriter::with_capacity(MAX_LINE_LENGTH, m.make_writer()),
+        )
+    })
+}
+
+struct LimitedWriter<W: Write>(usize, W);
```
Comment (tustvold) [inline trogging/src/lib.rs:444]: Why can't we wrap W with a Mutex?

Comment (mkmik) [reply trogging/src/lib.rs:444]: we don't pass `Write`rs to the subscriber but "writer factories", implementations of the `MakeWriter` trait.

The comment links to it: https://docs.rs/tracing-subscriber/0.2.18/tracing_subscriber/fmt/trait.MakeWriter.html

```rust
/// A type that can create [`io::Write`] instances.
///
/// `MakeWriter` is used by [`fmt::Subscriber`] or [`fmt::Layer`] to print formatted text representations of
/// [`Event`]s.
///
/// This trait is already implemented for function pointers and immutably-borrowing closures that
/// return an instance of [`io::Write`], such as [`io::stdout`] and [`io::stderr`].
....
pub trait MakeWriter {
    type Writer: Write;
    fn make_writer(&self) -> Self::Writer;
}
```

This trait is implemented for a Fn:

```rust
impl<F, W> MakeWriter for F
where
    F: Fn() -> W,
    W: io::Write,
{
    type Writer = W;

    fn make_writer(&self) -> Self::Writer {
        (self)()
    }
}
```

This is how the various example in `tracing-subscriber` work:

```rust
let subscriber = tracing_subscriber::registry()
    .with(fmt::Layer::new().with_writer(io::stdout));
```

This means, that every time `tracing` emits one log entry, it asks the "factory" to create a new Writer. In fact the docs say:

```
    /// # Implementer notes
    ///
    /// [`fmt::Layer`] or [`fmt::Subscriber`] will call this method each time an event is recorded. Ensure any state
    /// that must be saved across writes is not lost when the [`Writer`] instance is dropped. If
    /// creating a [`io::Write`] instance is expensive, be sure to cache it when implementing
    /// [`MakeWriter`] to improve performance.
```

---

On `HEAD`, tracing-subscriber now changed the trait as:

```rust
pub trait MakeWriter<'a> {
    type Writer: Write;
    fn make_writer(&'a self) -> Self::Writer;
}
```

This allows to implement it for `Mutex<W>`

```rust
impl<'a, W> MakeWriter<'a> for Mutex<W>
where
    W: io::Write + 'a,
{
    type Writer = MutexGuardWriter<'a, W>;

    fn make_writer(&'a self) -> Self::Writer {
        MutexGuardWriter(self.lock().expect("lock poisoned"))
    }
}
```

This lifetime is needed so we can express a mutex guard that lives less than the mutex it came from.

(Here "MutexGuardWriter" is just a plumbing that implements `Write` for a MutexGuard<Writer>)

---

So, I could have forked the library and cherrypicked that change; or I could have tried to compile against head (and see how many other breaking changes in the API there was).

That's why I said this was a hack; the right thing would be to fix it upstream (or to use the version of upstream that contains the proper solution.

Comment (mkmik) [reply trogging/src/lib.rs:444]: Relevant change in tracing-subscriber: https://github.com/tokio-rs/tracing/pull/781

**Comment by mkmik** [comment]
flaky test 

```
thread 'end_to_end_cases::persistence::test_full_lifecycle' panicked at 'Could not find chunk in desired state ObjectStoreOnly within 10s. Chunks were: [
```

File: trogging/src/lib.rs (hunk 1 above)
Comment (tustvold) [reply trogging/src/lib.rs:444]: Aah, thank you for explaining that for me. Makes sense

**Pull request merged**